File: rust/src/src/url_safety.rs
// src/url_safety.rs  <- src/url_safety.py
//! Outbound URL safety checks (SSRF hardening).
//!
//! Run before the server makes a request to a *user-supplied* URL — e.g. the
//! custom embedding endpoint set via `POST /api/embeddings/endpoint`, which then
//! triggers an outbound HTTP call.
//!
//! Odysseus is local-first: pointing an endpoint at a loopback or LAN address (a
//! local vLLM / llama.cpp / Ollama server) is a normal, intended setup, so this
//! guard does NOT blanket-block private addresses by default. What it ALWAYS
//! rejects: a non-HTTP(S) scheme, the link-local range (169.254.0.0/16 / fe80::/10
//! — the cloud instance-metadata SSRF vector), and multicast / reserved /
//! unspecified addresses. Set `EMBEDDING_BLOCK_PRIVATE_IPS=true` (`block_private`)
//! to additionally reject all private and loopback targets.

use std::net::{IpAddr, ToSocketAddrs};

const ALLOWED_SCHEMES: [&str; 2] = ["http", "https"];

/// `_default_resolver(host)` — resolve a hostname (or IP literal) to its IPs.
/// Python uses `socket.getaddrinfo(host, None)`; `ToSocketAddrs` with port 0 is
/// the stable analogue and likewise accepts IP literals verbatim.
fn default_resolver(host: &str) -> Vec<IpAddr> {
    match (host, 0u16).to_socket_addrs() {
        Ok(addrs) => addrs.map(|sa| sa.ip()).collect(),
        Err(_) => Vec::new(),
    }
}

/// `_classify(ip, block_private)` — rejection reason, or `None` if allowed.
fn classify(ip: IpAddr, block_private: bool) -> Option<String> {
    // IPv4-mapped IPv6 (e.g. ::ffff:169.254.169.254) -> judge the embedded v4.
    let ip = match ip {
        IpAddr::V6(v6) => match v6.to_ipv4_mapped() {
            Some(v4) => IpAddr::V4(v4),
            None => IpAddr::V6(v6),
        },
        v4 => v4,
    };
    let link_local = match ip {
        IpAddr::V4(a) => a.is_link_local(),
        // fe80::/10
        IpAddr::V6(a) => (a.segments()[0] & 0xffc0) == 0xfe80,
    };
    if link_local {
        return Some(format!("link-local address blocked (SSRF metadata risk): {ip}"));
    }
    let reserved = match ip {
        IpAddr::V4(a) => a.octets()[0] >= 240, // 240.0.0.0/4 (is_reserved is unstable)
        IpAddr::V6(_) => false,
    };
    if ip.is_multicast() || reserved || ip.is_unspecified() {
        return Some(format!("disallowed address: {ip}"));
    }
    if block_private {
        let private = match ip {
            IpAddr::V4(a) => a.is_private(),
            // fc00::/7 (unique-local).
            IpAddr::V6(a) => matches!(a.segments()[0] >> 8, 0xfc | 0xfd),
        };
        if private || ip.is_loopback() {
            return Some(format!("private/loopback address blocked: {ip}"));
        }
    }
    None
}

/// `check_outbound_url(url, block_private=False)` — validate a user-supplied
/// outbound URL. Returns `(ok, reason)`; `ok` is true only when safe to fetch.
pub fn check_outbound_url(url: &str, block_private: bool) -> (bool, String) {
    check_outbound_url_with(url, block_private, default_resolver)
}

/// As [`check_outbound_url`] but with an injectable resolver (the Python
/// `resolver=` kwarg) so tests can avoid real DNS.
pub fn check_outbound_url_with(
    url: &str,
    block_private: bool,
    resolver: impl Fn(&str) -> Vec<IpAddr>,
) -> (bool, String) {
    let trimmed = url.trim();
    if trimmed.is_empty() {
        return (false, "URL is required".to_string());
    }
    let parsed = match url::Url::parse(trimmed) {
        Ok(p) => p,
        Err(e) => return (false, format!("unparseable URL: {e}")),
    };
    let scheme = parsed.scheme().to_lowercase();
    if !ALLOWED_SCHEMES.contains(&scheme.as_str()) {
        let shown = if scheme.is_empty() { "(none)" } else { scheme.as_str() };
        return (false, format!("scheme must be http or https, got '{shown}'"));
    }
    // urlparse(...).hostname strips IPv6 brackets; mirror that for the resolver.
    let host = match parsed.host_str() {
        Some(h) if !h.is_empty() => h.trim_start_matches('[').trim_end_matches(']').to_string(),
        _ => return (false, "URL has no host".to_string()),
    };
    let ips = resolver(&host);
    if ips.is_empty() {
        return (false, "host does not resolve".to_string());
    }
    for ip in ips {
        if let Some(reason) = classify(ip, block_private) {
            return (false, reason);
        }
    }
    (true, "ok".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::net::Ipv4Addr;

    fn r(ip: &str) -> impl Fn(&str) -> Vec<IpAddr> + '_ {
        let ip = ip.to_string();
        move |_| vec![ip.parse().unwrap()]
    }

    #[test]
    fn rejects_non_http_scheme_and_empty() {
        assert!(!check_outbound_url("", false).0);
        assert!(!check_outbound_url_with("file:///etc/passwd", false, r("1.1.1.1")).0);
        assert!(!check_outbound_url_with("ftp://h/x", false, r("1.1.1.1")).0);
    }

    #[test]
    fn always_blocks_link_local_metadata() {
        // 169.254.169.254 (cloud metadata) is rejected even with block_private=false.
        let (ok, reason) = check_outbound_url_with("http://meta/", false, r("169.254.169.254"));
        assert!(!ok);
        assert!(reason.contains("link-local"));
        // IPv4-mapped form too.
        assert!(!check_outbound_url_with("http://meta/", false, r("::ffff:169.254.169.254")).0);
    }

    #[test]
    fn public_allowed_private_gated() {
        assert!(check_outbound_url_with("http://api/", false, r("1.1.1.1")).0);
        // private is allowed by default (local-first) ...
        assert!(check_outbound_url_with("http://lan/", false, r("192.168.1.5")).0);
        // ... but rejected when block_private is set.
        assert!(!check_outbound_url_with("http://lan/", true, r("192.168.1.5")).0);
        assert!(!check_outbound_url_with("http://lo/", true, r("127.0.0.1")).0);
        let _ = Ipv4Addr::LOCALHOST;
    }
}

File: src/url_safety.py
"""Outbound URL safety checks (SSRF hardening).

Run before the server makes a request to a *user-supplied* URL — e.g. the custom
embedding endpoint set via ``POST /api/embeddings/endpoint``, which then triggers
an outbound ``httpx`` call.

Odysseus is local-first: pointing the embedding endpoint at a loopback or LAN
address (a local vLLM / llama.cpp / Ollama server) is a normal, intended setup.
So this guard does **not** blanket-block private addresses by default — that would
break the primary use case. What it *always* rejects:

  - a non-HTTP(S) scheme (``file://``, ``gopher://``, ``ftp://`` …), and
  - the link-local range (``169.254.0.0/16`` / ``fe80::/10``), i.e. the cloud
    instance-metadata SSRF credential-exfil vector — nobody serves embeddings
    there — plus multicast / reserved / unspecified addresses.

For exposed multi-tenant deployments, set ``EMBEDDING_BLOCK_PRIVATE_IPS=true`` to
additionally reject all private and loopback targets (full SSRF lockdown).
"""

import ipaddress
import socket
from typing import Callable, List, Optional, Tuple
from urllib.parse import urlparse

ALLOWED_SCHEMES = ("http", "https")

# RFC 6598 shared address space (carrier-grade NAT). It is not globally
# routable, but CPython does not classify it as ``is_private`` (it is "shared",
# not "private"), so the is_private/is_loopback checks miss it. Reject the range
# explicitly. This closes exactly the shared-space gap without coupling strict
# mode to ``is_global``'s broader definition, which has shifted across CPython
# versions for other special ranges.
_SHARED_ADDRESS_SPACE_V4 = ipaddress.ip_network("100.64.0.0/10")


def _default_resolver(host: str) -> List[str]:
    """Resolve a hostname to the list of IP strings it maps to (A + AAAA)."""
    return [info[4][0] for info in socket.getaddrinfo(host, None)]


def _classify(ip: ipaddress._BaseAddress, *, block_private: bool) -> Optional[str]:
    """Return a rejection reason for an IP, or None if it is allowed."""
    # IPv4-mapped IPv6 (e.g. ::ffff:169.254.169.254) — judge the embedded v4.
    if isinstance(ip, ipaddress.IPv6Address) and ip.ipv4_mapped is not None:
        ip = ip.ipv4_mapped
    if ip.is_link_local:
        return f"link-local address blocked (SSRF metadata risk): {ip}"
    if ip.is_multicast or ip.is_reserved or ip.is_unspecified:
        return f"disallowed address: {ip}"
    if block_private and (
        ip.is_private
        or ip.is_loopback
        or (isinstance(ip, ipaddress.IPv4Address) and ip in _SHARED_ADDRESS_SPACE_V4)
    ):
        return f"private/shared/loopback address blocked: {ip}"
    return None


def check_outbound_url(
    url: str,
    *,
    block_private: bool = False,
    resolver: Optional[Callable[[str], List[str]]] = None,
) -> Tuple[bool, str]:
    """Validate a user-supplied outbound URL.

    Returns ``(ok, reason)``. ``ok`` is True only when the URL is safe to fetch.
    ``resolver`` is injectable so callers/tests can avoid real DNS.
    """
    if not isinstance(url, str):
        return False, "URL must be a string"
    if not url or not url.strip():
        return False, "URL is required"
    try:
        parsed = urlparse(url.strip())
    except Exception as e:  # pragma: no cover - urlparse is very tolerant
        return False, f"unparseable URL: {e}"

    if parsed.scheme.lower() not in ALLOWED_SCHEMES:
        return False, f"scheme must be http or https, got '{parsed.scheme or '(none)'}'"
    host = parsed.hostname
    if not host:
        return False, "URL has no host"

    resolve = resolver or _default_resolver
    try:
        raw_ips = resolve(host)
    except Exception as e:
        return False, f"host does not resolve: {e}"
    if not raw_ips:
        return False, "host does not resolve"

    saw_ip = False
    for raw in raw_ips:
        if not isinstance(raw, str):
            continue
        try:
            ip = ipaddress.ip_address(raw.split("%")[0])  # strip IPv6 zone id
        except ValueError:
            continue
        saw_ip = True
        reason = _classify(ip, block_private=block_private)
        if reason:
            return False, reason
    if not saw_ip:
        return False, "host does not resolve to an IP"
    return True, "ok"


_LOCALHOST_NAMES = ("localhost", "ip6-localhost", "ip6-loopback")


def _is_localhost_name(host: str) -> bool:
    h = host.lower().rstrip(".")
    return h in _LOCALHOST_NAMES or h.endswith(".localhost")


def check_outbound_host(
    host: str,
    *,
    block_private: bool = False,
    resolver: Optional[Callable[[str], List[str]]] = None,
    unresolved: str = "reject",
) -> Tuple[bool, str]:
    """Validate a bare user-supplied hostname / IP literal (no scheme).

    Companion to :func:`check_outbound_url` for protocols that take a host
    and port rather than a URL — IMAP, SMTP, CalDAV/CardDAV hosts, database
    DSNs. Same policy: link-local / multicast / reserved / unspecified
    addresses are always rejected (cloud metadata SSRF), private / loopback
    only when ``block_private`` is set. ``localhost`` spellings are treated as
    loopback without a DNS round-trip.

    ``unresolved`` controls what happens when a hostname does not resolve:

    - ``"reject"`` (default, fail-closed like ``check_outbound_url``);
    - ``"allow"``: pass it through. Use this only where the very next step
      is a connect that will surface the same DNS failure to the user, so
      an offline host or a flaky resolver does not turn into a spurious
      "blocked" error. IP literals and localhost names are still classified
      without DNS, so the metadata / loopback cases are covered either way.
    """
    if not isinstance(host, str):
        return False, "host must be a string"
    host = host.strip()
    if not host:
        return False, "host is required"
    # A bare host never carries a scheme, path, userinfo, port or whitespace.
    # Anything of the sort is either a copy/paste mistake or an attempt to
    # smuggle a target past the check; reject rather than guess.
    if any(ch.isspace() for ch in host) or any(ch in host for ch in "/\\@?#"):
        return False, "host must be a bare hostname or IP address"
    literal = host[1:-1] if host.startswith("[") and host.endswith("]") else host
    if ":" in literal:
        # Only an IPv6 literal may contain colons; "host:port" is rejected.
        try:
            ip = ipaddress.ip_address(literal.split("%")[0])
        except ValueError:
            return False, "host must be a bare hostname or IP address"
        reason = _classify(ip, block_private=block_private)
        return (False, reason) if reason else (True, "ok")
    try:
        ip = ipaddress.ip_address(literal)
    except ValueError:
        ip = None
    if ip is not None:
        reason = _classify(ip, block_private=block_private)
        return (False, reason) if reason else (True, "ok")
    if _is_localhost_name(host):
        if block_private:
            return False, f"private/shared/loopback address blocked: {host}"
        return True, "ok"

    resolve = resolver or _default_resolver
    try:
        raw_ips = resolve(host)
    except Exception as e:
        if unresolved == "allow":
            return True, "unresolved (deferred to connect)"
        return False, f"host does not resolve: {e}"
    saw_ip = False
    for raw in raw_ips or []:
        if not isinstance(raw, str):
            continue
        try:
            resolved = ipaddress.ip_address(raw.split("%")[0])
        except ValueError:
            continue
        saw_ip = True
        reason = _classify(resolved, block_private=block_private)
        if reason:
            return False, reason
    if not saw_ip:
        if unresolved == "allow":
            return True, "unresolved (deferred to connect)"
        return False, "host does not resolve to an IP"
    return True, "ok"
